### PR TITLE
Kotlin M13 support.

### LINKED ISSUE #19: Kotlin m13

### PR DESCRIPTION
**Pull request by yoavst**
I've converted the project to Kotlin M13. (fixes #19 )
In some places it uses `lazy(LazyThreadSafetyMode.NONE)` instead of just `lazy`. That is because the automatic fix on all the project. I think it works the same as `lazy`, so you can replace it later.
Also, in order to remove a warning, I've added `Method.PATCH` to the when expression. However, `Method.PATCH` is inaccessible as a method, so it doesn't actually change something.

```kotlin
private fun setDoOutput(connection: HttpURLConnection, method: Method) {
         when (method) {
            Method.GET, Method.DELETE -> connection.doOutput = false
            Method.POST, Method.PUT, Method.PATCH -> connection.doOutput = true
         }
     }
```

Also I've run the tests, all pass. So it looks like no functionality were changed.

### CODE REVIEW

**Comment by kittinunf** [comment]
@yoav. That's awesome!!! I was skimming through changes and everything is good! :+1. Let me build on my machine once again and I ll merge this in asap. 



—

File: fuel/src/main/kotlin/fuel/core/Request.kt [hunk 1]
```diff
@@ -189,9 +189,9 @@ public class Request {
     public fun responseJson(handler: Handler<JSONObject>): Unit = response(Request.jsonDeserializer(), handler)
 
     //object
-    public fun <T> responseObject(deserializer: ResponseDeserializable<T>, handler: (Request, Response, Either<FuelError, T>) -> Unit): Unit = response(deserializer, handler)
+    public fun <T: Any> responseObject(deserializer: ResponseDeserializable<T>, handler: (Request, Response, Either<FuelError, T>) -> Unit): Unit = response(deserializer, handler)
```
Comment (kittinunf) [inline fuel/src/main/kotlin/fuel/core/Request.kt]: Do you have a specific reason why we need to give `T` an upper bound with `Any`?

**Comment by kittinunf** [comment]
Document for `lazy` says 

``` Kotlin
/**
 * Creates a new instance of the [Lazy] that uses the specified initialization function [initializer]
 * and the default thread-safety mode [LazyThreadSafetyMode.SYNCHRONIZED].
**/
```

So it seems like plain `lazy` would automatically be synchronized. Meanwhile, the `lazy(LazyThreadSafetyMode.NONE)` wouldn't.  

Perhaps, just to be safe, I guess we should stick with the non-synchronized version as I don't think it is needed.

**Comment by yoavst** [comment]
I didn't know about the lazy. 

About generics.

Maybe @mplatvoet would be able to explain it better since he is the developer of [Kovenant](https://github.com/mplatvoet/kovenant).

IDEA's Kotlin M13 plugin shows a warning if `T` doesn't extend `any` when you use `T?`.
 If the methods I changed can return `T?`, then you better keep it `T : Any` but change the return type to `T?`

**Comment by kittinunf** [comment]
OK. I think you are right.

BTW, I think I will create a proper branch for this because I would like to fix #21 created by him. Would you mind create another PR against my 0.53 branch? Then I would fix #21 together with it.

**Comment by yoavst** [comment]
I'm not a git pro, so I'll have to reapply all the changes. I can do another commit with the sealed change, lazy fix, and nullability fix for deserializer if needed.

**Comment by kittinunf** [comment]
I created this for you to merge your M13 fix in. #22

**Comment by kittinunf** [comment]
C'mon .. I would love to not let you do this alone. Lemme help :) I will fix other small things as well.

**Comment by mplatvoet** [comment]
I think the lazy instance *should* be thread safe in most cases here. Same goes for `readWriteLazy` which is used for instance in headers. The resulting Request instance can be shared amongst threads and I think people will. Only if you can guarantee that a lazy can only be initialised and read by one thread you should use the unsynced variant.

**Comment by yoavst** [comment]
Also, what about nullability in generics for `Deserializable`?
Should it be able to return null?

**Pull request closed without merging**